Web Inspector: Console code completion doesn't suggest variables unless inspector is re-opened

the-chenergy opened the pull request:
#### 5c6736f3f9ad8508918c6a2c2ece135498603d9c
<pre>
Web Inspector: Console code completion doesn&apos;t suggest variables unless inspector is re-opened
<a href="https://rdar.apple.com/125035133">rdar://125035133</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=271265">https://bugs.webkit.org/show_bug.cgi?id=271265</a>

Reviewed by Devin Rousso.

In the current JavaScriptRuntimeCompletionProvider, there&apos;s work done
to cache the list of property names present in the runtime. This
property name cache is cleared / forced to be re-fetched either when 30
seconds have passed or when the code typed starts accessing a different
object&apos;s properties (the object to the left of a dot operator). However,
when variables are created by executing a command, the cache is not
updated, resulting in those new variables missing from the suggestions.

This commit simply makes it so that executing a command from the console
also forces clearing the cache, so that new variables, in case created,
are immediately collected as suggestions when typing a new command. The
cache is also cleared when the console clears, so when the page is
reloaded with the console opened, the variable completion still gets
properly reset.

* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
(WI.JavaScriptRuntimeCompletionProvider):
(WI.JavaScriptRuntimeCompletionProvider.prototype.completionControllerCompletionsNeeded):
   - Update the comment on the need of loading the list of property
     names from cache to reflect now there can be additional
     opportunities that the cache should be cleared.
   - Remove the `// Protected` section comment as this function is
     accessed by others.

(WI.JavaScriptRuntimeCompletionProvider):
(WI.JavaScriptRuntimeCompletionProvider.prototype.clearCachedPropertyNames):
(WI.JavaScriptRuntimeCompletionProvider.prototype._clearLastProperties): Deleted
   - Expose the logic of clearing the property name cache so that the
     completion provider&apos;s users can also force a reload of property
     names.

(WI.JavaScriptRuntimeCompletionProvider.prototype.completionControllerCompletionsNeeded.updateLastPropertyNames):
   - Adapt to the cache-clearing logic being exposed to public.

* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
(WI.JavaScriptLogViewController.prototype.consolePromptTextCommitted):
   - Clear cached property names whenever a new command is executed in
     the console, since new commands are what can possibly introduce
     changes to existing variables (additions from `x = ...` or `let x`,
     and deletions from `delete x`).

* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager.prototype.messagesCleared):
   - Clear cached property names when the console is cleared because the
     page might&apos;ve been reloaded and therefore variables were reset
     as a result.

* Source/WebInspectorUI/UserInterface/Test.html:
* LayoutTests/inspector/console/js-completions.html: Added.
* LayoutTests/inspector/console/js-completions-expected.txt: Added.
   - Add a simple test to demonstrate how `clearCachedPropertyNames`
     is used.

Canonical link: <a href="https://commits.webkit.org/277261@main">https://commits.webkit.org/277261@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b0a0b42b1c43fc2f022d555c341a1e680821edd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47152 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26331 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49835 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43201 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31641 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23788 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38410 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47733 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40634 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19719 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41786 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5196 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42193 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51711 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22177 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18547 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45704 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23454 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/40779 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44712 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10397 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24236 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23171 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

the-chenergy commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -160,9 +174,9 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
-            this._clearLastPropertiesTimeout = setTimeout(this._clearLastProperties.bind(this), WI.JavaScriptLogViewController.CachedPropertiesDuration);
+            if (this.clearCacheTimeout !== undefined)
+                clearTimeout(this.clearCacheTimeout);
+            this.clearCacheTimeout = setTimeout(this.clearCache.bind(this), WI.JavaScriptLogViewController.CachedPropertiesDuration);
```
I wonder if this `setTimeout` thing is still useful. What was the purpose for reloading every 30 seconds? 🤔 @dcrousso

the-chenergy commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -332,7 +346,6 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
                 }
 
-                // FIXME: Due to caching, sometimes old $n values show up as completion results even though they are not available. We should clear that proactively.
                 for (var i = 1; i <= WI.ConsoleCommandResultMessage.maximumSavedResultIndex; ++i) {
```
Would the code be a bit prettier if we avoid reading the global-ish variable `maximumSavedResultIndex` from here, but instead let the variable be an instance variable here and let `ConsoleCommandResultMessage` reach into this `WI.javaScriptRuntimeCompletionProvider` to modify it instead? @dcrousso 

If we think that'd be an improvement, I can make a bug report to do that next.

Otherwise, my way of reaching into this completion provider object to clear its cache feels a little inconsistent with what this piece of code is doing (reaching elsewhere).

dcrousso replied:
are we sure that this FIXME is no longer valid?  i can imagine a number of scenarios where it's still possible to have invalid cached property names

francescorn replied:
If the FIXME has been tested in the mini-browser, we can delete the FIXME. If not, we should probably keep it.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -91,6 +91,20 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
+    {
+        if (this.clearCacheTimeout) {
+            clearTimeout(this.clearCacheTimeout);
+            delete this.clearCacheTimeout;
```
we prefer avoiding `delete` as it can cause the underlying JSC structure to change for this object which can be a deoptimization.  please just set it to `undefined` instead
```js
this._clearCacheTimeout = undefined;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -160,9 +174,9 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
-            if (this._clearLastPropertiesTimeout)
-                clearTimeout(this._clearLastPropertiesTimeout);
-            this._clearLastPropertiesTimeout = setTimeout(this._clearLastProperties.bind(this), WI.JavaScriptLogViewController.CachedPropertiesDuration);
+            if (this.clearCacheTimeout !== undefined)
```
this should be "private" `_clearCacheTimeout` since it's not used outside of this object

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -91,6 +91,20 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
 
+    // Public
+
+    clearCache()
```
i think this needs a more specific name like `clearCachedPropertyNames` or something so that it's more clear what it's doing

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ConsolePrompt.js:
```diff
@@ -267,6 +267,8 @@ WI.ConsolePrompt = class ConsolePrompt extends WI.View
             if (this.delegate && typeof this.delegate.consolePromptTextCommitted === "function")
                 this.delegate.consolePromptTextCommitted(this, currentText);
+
+            WI.javaScriptRuntimeCompletionProvider.clearCache();
```
this is not the right place to put this, as currently `WI.ConsolePrompt` is used for more than just JavaScript (e.g. WebSQL)

i think this should be moved to `WI.JavaScriptLogViewController.prototype.consolePromptTextCommitted`

dcrousso reviewed (commented):
great catch!  a few minor adjustments/fixes, but overall well done :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -91,6 +91,19 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
 
+    // Public
+
+    // Clears the cache of property names to force picking up any changes occurred while stepping, executing new code in the console prompt, or sitting idle.
```
Style: the comment should be inside the function

also, most of this depends more on how it's used, not this code itself, so if you're determined to keep the comment id probably move it to the various callsites instead (and my preference would be to drop it entirely since the name of the function should be clear enough to explain what it's for, and if not we should make the name clearer)

the-chenergy commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -160,9 +172,11 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
+            if (this._clearCachedPropertyNamesTimeout)
+                clearTimeout(this._clearCachedPropertyNamesTimeout);
+            // Clears the cache of property names to force picking up any changes occurred while stepping, executing new code in the console prompt, or sitting idle.
+            // 
```
Whoops, I think I meant to create a bug and write a FIXME here about the 30-second thing

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -264,6 +264,8 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
 
         WI.runtimeManager.evaluateInInspectedWindow(text, options, printResult.bind(this));
+
+        WI.javaScriptRuntimeCompletionProvider.clearCachedPropertyNames();
```
NIT: from an order of operations perspective, i think it'd be safer to clear the cached property names before `WI.runtimeManager.evaluateInInspectedWindow` just in case it does something that'd affect the cached property names (which i realize is not the case today, but could change in the future and result in a potentially very subtle bug)

dcrousso reviewed (approved):
nice work!  thanks for iterating with me :)

francescorn commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
```diff
@@ -91,7 +91,17 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
+
+        this._lastPropertyNames = null;
+    }
 
```
NIT: Add a `// Protected` here unless we are making the rest public.

the-chenergy replied:
Ah yes, `clearCachedPropertyNames()` was meant to be public, and I removed the `// Protected` comment since the function below that it used to mark as protected, `completionControllerCompletionsNeeded()`, is treated as public and called by others anyway (such as [`CodeMirrorCompletionController`](https://github.com/WebKit/WebKit/blob/b3eceeb22f58375c3ae39ccdc376335e417881fe/Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js#L533)), so I thought marking it as protected would be misleading 😊

rcaliman-apple commented on LayoutTests/inspector/console/js-completions.html:
```diff
@@ -0,0 +1,54 @@
+            InspectorTest.expectThat(!completions.includes("myVariable"), "Completions should not contain myVariable by default.");
+
+            await InspectorTest.evaluateInPage(`myVariable = 42;`);
+            WI.javaScriptRuntimeCompletionProvider.clearCachedPropertyNames();
```
I'm a slightly concerned that clearing the cache manually will obscure a future bug where the implementation breaks but the test still passes.

Could we instead set `WI.JavaScriptLogViewController.CachedPropertiesDuration = 0` at the beginning of the test and then reset it at the end?

I'm thinking that accelerating time will still test the implementation behavior, but I am concerned that this will introduce flakiness. Pinging @dcrousso for a second opinion.

the-chenergy replied:
Ah, I too was trying to come up with a way to let the internal code to clear the cache by itself without having to call it here, but I couldn't think of any. Clearing the cache here manually does look a bit fake.

Is there way for a test to mock running code in inspector's console? (`InspectorTest.evaluateInPage` and `TestPage.evaluateInWebInspector` both don't actually mimic running the code in console and therefore don't trigger `JavaScriptLogViewController.consolePromptTextCommitted`. I thought it'd be best if there's a way to run code that does trigger the said function.)

> Could we instead set `WI.JavaScriptLogViewController.CachedPropertiesDuration = 0` at the beginning of the test and then reset it at the end?

That's an interesting idea! But that's then relying on the completion provider's auto-clearing procedure, still not the console proactively clearing the cache when new code is run, which is the behavior my patch was actually adding 🤔

rcaliman-apple replied:
Web Inspector does not have UI tests. If the completions test requires Views to be instantiated or any UI interaction, we can't really test that.

If that's the case, clearing the cache manually to test the controller behavior is fine.

the-chenergy replied:
OK.

In that case, I think I'd wrap the cache-clearing call into a mock function like this:
```suggestion
            const mockHandleConsolePromptCommitted = () => {
                WI.javaScriptRuntimeCompletionProvider.clearCachedPropertyNames();
            };
            mockHandleConsolePromptCommitted();
```

Just to be clear that this cache-clearing call is supposed to be done by somebody else.

rcaliman-apple replied:
It's fine to keep the method inline. Having clear and simple to read tests is good ;)

dcrousso replied:
FWIW this test is more about verifying that `WI.javaScriptRuntimeCompletionProvider` works as intended, not necessarily that the hookups to console evaluations are working

as such, id maybe even add some negative tests to verify that things are still (not) in the cache after evaluating until the cache is cleared
```js
let completionsBefore = await generateJSCompletions();
InspectorTest.expectThat(!completions.includes("myVariable"), "Completions should not contain myVariable by default.");

await InspectorTest.evaluateInPage(`myVariable = 42;`);
let completionsAfterEvaluating = await generateJSCompletions();
InspectorTest.expectThat(!completionsAfterEvaluating.includes("myVariable"), "Completions should not contain myVariable after evaluation unless the cache is cleared.");
InspectorTest.assert(Object.shallowEqual(completionsBefore, completionsAfterEvaluating), "Completions should not change unless the cache is cleared.");

WI.javaScriptRuntimeCompletionProvider.clearCachedPropertyNames();
let completionsAfterClearingForEvaluation = await generateJSCompletions();
InspectorTest.expectThat(completionsAfterClearingForEvaluation.includes("myVariable"), "Completions should contain myVariable after it's created by code run in console.");
InspectorTest.assert(!Object.shallowEqual(completionsAfterEvaluating, completionsAfterClearingForEvaluation), "Completions should change after the cache is cleared.");

await InspectorTest.evaluateInPage(`delete myVariable;`);
let completionsAfterDeleting = await generateJSCompletions();
InspectorTest.expectThat(completionsAfterDeleting.includes("myVariable"), "Completions should contain myVariable after deleting unless the cache is cleared.");
InspectorTest.assert(Object.shallowEqual(completionsAfterClearingForEvaluation, completionsAfterDeleting), "Completions should not change unless the cache is cleared.");

WI.javaScriptRuntimeCompletionProvider.clearCachedPropertyNames();
let completionsAfterClearingForDelete = await generateJSCompletions();
InspectorTest.expectThat(!completionsAfterClearingForDelete.includes("myVariable"), "Completions should not contain myVariable after it's deleted by code run in console.");
InspectorTest.assert(!Object.shallowEqual(completionsAfterDeleting, completionsAfterClearingForDelete), "Completions should change after the cache is cleared.");
```

the-chenergy committed:
Web Inspector: Console code completion doesn't suggest variables unless inspector is re-opened
rdar://125035133
https://bugs.webkit.org/show_bug.cgi?id=271265

Reviewed by Devin Rousso.

In the current JavaScriptRuntimeCompletionProvider, there's work done
to cache the list of property names present in the runtime. This
property name cache is cleared / forced to be re-fetched either when 30
seconds have passed or when the code typed starts accessing a different
object's properties (the object to the left of a dot operator). However,
when variables are created by executing a command, the cache is not
updated, resulting in those new variables missing from the suggestions.

This commit simply makes it so that executing a command from the console
also forces clearing the cache, so that new variables, in case created,
are immediately collected as suggestions when typing a new command. The
cache is also cleared when the console clears, so when the page is
reloaded with the console opened, the variable completion still gets
properly reset.

* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
(WI.JavaScriptRuntimeCompletionProvider):
(WI.JavaScriptRuntimeCompletionProvider.prototype.completionControllerCompletionsNeeded):
   - Update the comment on the need of loading the list of property
     names from cache to reflect now there can be additional
     opportunities that the cache should be cleared.
   - Remove the `// Protected` section comment as this function is
     accessed by others.

(WI.JavaScriptRuntimeCompletionProvider):
(WI.JavaScriptRuntimeCompletionProvider.prototype.clearCachedPropertyNames):
(WI.JavaScriptRuntimeCompletionProvider.prototype._clearLastProperties): Deleted
   - Expose the logic of clearing the property name cache so that the
     completion provider's users can also force a reload of property
     names.

(WI.JavaScriptRuntimeCompletionProvider.prototype.completionControllerCompletionsNeeded.updateLastPropertyNames):
   - Adapt to the cache-clearing logic being exposed to public.

* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
(WI.JavaScriptLogViewController.prototype.consolePromptTextCommitted):
   - Clear cached property names whenever a new command is executed in
     the console, since new commands are what can possibly introduce
     changes to existing variables (additions from `x = ...` or `let x`,
     and deletions from `delete x`).

* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager.prototype.messagesCleared):
   - Clear cached property names when the console is cleared because the
     page might've been reloaded and therefore variables were reset
     as a result.

* Source/WebInspectorUI/UserInterface/Test.html:
* LayoutTests/inspector/console/js-completions.html: Added.
* LayoutTests/inspector/console/js-completions-expected.txt: Added.
   - Add a simple test to demonstrate how `clearCachedPropertyNames`
     is used.

Canonical link: https://commits.webkit.org/277261@main
